File: lib/NUnit.org/NUnit/2.5.9/samples/csharp/money/Money.cs
// ****************************************************************
// This is free software licensed under the NUnit license. You
// may obtain a copy of the license as well as information regarding
// copyright ownership at http://nunit.org/?p=license&r=2.4.
// ****************************************************************

namespace NUnit.Samples.Money 
{

	using System;
	using System.Text;

	/// <summary>A simple Money.</summary>
	class Money: IMoney 
	{

		private int fAmount;
		private String fCurrency;
        
		/// <summary>Constructs a money from the given amount and
		/// currency.</summary>
		public Money(int amount, String currency) 
		{
			fAmount= amount;
			fCurrency= currency;
		}

		/// <summary>Adds a money to this money. Forwards the request to
		/// the AddMoney helper.</summary>
		public IMoney Add(IMoney m) 
		{
			return m.AddMoney(this);
		}

		public IMoney AddMoney(Money m) 
		{
			if (m.Currency.Equals(Currency) )
				return new Money(Amount+m.Amount, Currency);
			return new MoneyBag(this, m);
		}

		public IMoney AddMoneyBag(MoneyBag s) 
		{
			return s.AddMoney(this);
		}

		public int Amount 
		{
			get { return fAmount; }
		}

		public String Currency 
		{
			get { return fCurrency; }
		}

		public override bool Equals(Object anObject) 
		{
			if (IsZero)
				if (anObject is IMoney)
					return ((IMoney)anObject).IsZero;
			if (anObject is Money) 
			{
				Money aMoney= (Money)anObject;
				return aMoney.Currency.Equals(Currency)
					&& Amount == aMoney.Amount;
			}
			return false;
		}

		public override int GetHashCode() 
		{
			return fCurrency.GetHashCode()+fAmount;
		}

		public bool IsZero 
		{
			get { return Amount == 0; }
		}

		public IMoney Multiply(int factor) 
		{
			return new Money(Amount*factor, Currency);
		}

		public IMoney Negate() 
		{
			return new Money(-Amount, Currency);
		}

		public IMoney Subtract(IMoney m) 
		{
			return Add(m.Negate());
		}

		public override String ToString() 
		{
			StringBuilder buffer = new StringBuilder();
			buffer.Append("["+Amount+" "+Currency+"]");
			return buffer.ToString();
		}
	}
}

File: tags/Lucene.Net_2_9_4_RC2/lib/NUnit.org/NUnit/2.5.9/samples/vb/money/Money.vb
' ****************************************************************
' This is free software licensed under the NUnit license. You
' may obtain a copy of the license as well as information regarding
' copyright ownership at http://nunit.org/?p=license&r=2.4.
' ****************************************************************

Option Explicit On 

Namespace NUnit.Samples

    ' A Simple Money.
    Public Class Money
        Implements IMoney

        Private fAmount As Int32
        Private fCurrency As String

        ' Constructs a money from a given amount and currency.
        Public Sub New(ByVal amount As Int32, ByVal currency As String)
            Me.fAmount = amount
            Me.fCurrency = currency
        End Sub


        ' Adds a money to this money. Forwards the request
        ' to the AddMoney helper.
        Public Overloads Function Add(ByVal m As IMoney) As IMoney Implements IMoney.Add
            Return m.AddMoney(Me)
        End Function

        Public Overloads Function AddMoney(ByVal m As Money) As IMoney Implements IMoney.AddMoney
            If m.Currency.Equals(Currency) Then
                Return New Money(Amount + m.Amount, Currency)
            End If

            Return New MoneyBag(Me, m)
        End Function

        Public Function AddMoneyBag(ByVal s As MoneyBag) As IMoney Implements IMoney.AddMoneyBag
            Return s.AddMoney(Me)
        End Function

        Public ReadOnly Property Amount() As Integer
            Get
                Return fAmount
            End Get
        End Property

        Public ReadOnly Property Currency() As String
            Get
                Return fCurrency
            End Get
        End Property

        Public Overloads Overrides Function Equals(ByVal anObject As Object) As Boolean
            If IsZero And TypeOf anObject Is IMoney Then
                Dim aMoney As IMoney = anObject
                Return aMoney.IsZero
            End If

            If TypeOf anObject Is Money Then
                Dim aMoney As Money = anObject
                If (IsZero) Then
                    Return aMoney.IsZero
                End If

                Return Currency.Equals(aMoney.Currency) And Amount.Equals(aMoney.Amount)
            End If

            Return False
        End Function

        Public Overrides Function GetHashCode() As Int32
            Return fCurrency.GetHashCode() + fAmount
        End Function

        Public ReadOnly Property IsZero() As Boolean Implements IMoney.IsZero
            Get
                Return Amount.Equals(0)
            End Get
        End Property

        Public Function Multiply(ByVal factor As Integer) As IMoney Implements IMoney.Multiply

            Return New Money(Amount * factor, Currency)

        End Function

        Public Function Negate() As IMoney Implements IMoney.Negate

            Return New Money(-Amount, Currency)

        End Function

        Public Function Subtract(ByVal m As IMoney) As IMoney Implements IMoney.Subtract

            Return Add(m.Negate())

        End Function

        Public Overrides Function ToString() As String

            Return String.Format("[{0} {1}]", Amount, Currency)

        End Function

    End Class

End Namespace
